dataflow: extensible template with javascript udf

anguillanneuf opened the pull request:
Toward internal b/238353202 and b/244173472 (region tags are created on Monday, Aug 29)

dandhlee commented on dataflow/conftest.py:
```diff
@@ -525,6 +555,46 @@ def dataflow_flex_template_run(
+                **parameters,
+            }.items()
+        ]
+        logging.info(f"{cmd}")
```
```suggestion
        logging.info(cmd)
```

dandhlee reviewed (commented):
Will you be adding the region tag in the future?

anguillanneuf commented:
@dandhlee Oh yes, region tags were created (this morning) and added.

dandhlee reviewed (approved):
LGTM for Python.
Could you get a review for the JS file if you think it's needed? If not, just a product review would be fine.

davidcavazos commented on dataflow/conftest.py:
```diff
@@ -91,23 +92,50 @@ def storage_bucket(name: str) -> str:
 
-        logging.info(f"Created bigquery_dataset: {dataset.full_dataset_id}")
-        yield dataset_name
+        schema = [
```
Can we make this `bigquery_table` to be a general purpose wrapper? How about we make the schema an optional parameter?

davidcavazos commented on dataflow/conftest.py:
```diff
@@ -91,23 +92,50 @@ def storage_bucket(name: str) -> str:
-        logging.info(f"Deleted bigquery_dataset: {dataset.full_dataset_id}")
+        result = bigquery_client.create_table(table)
+        logging.info(f"Created bigquery_table: {result.full_table_id}")
+        yield result.table_id
```
Should we be deleting the table here as well? Or are we assuming the table being deleted when the dataset is deleted? If so, we should add a comment.

davidcavazos commented on dataflow/conftest.py:
```diff
@@ -525,6 +555,46 @@ def dataflow_flex_template_run(
         Utils.dataflow_jobs_cancel(job_id)
 
+    @staticmethod
+    def dataflow_extensible_template_run(
```
I think the "official" name for the older (non-Flex) templates is "Classic Templates". Can we rename this to `dataflow_classic_template_run`? Or was there a reason why "extensible"?

https://cloud.google.com/dataflow/docs/guides/templates/creating-templates

davidcavazos commented on dataflow/extensible-templates/dataflow_udf_transform.js:
```diff
@@ -0,0 +1,35 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START dataflow_extensible_template_udf]
```
If we decide to rename from `extensible` to `classic` we might want to change this as well.

davidcavazos commented on dataflow/extensible-templates/noxfile_config.py:
```diff
@@ -0,0 +1,42 @@
+
+TEST_CONFIG_OVERRIDE = {
+    # You can opt out from the test for specific Python versions.
+    "ignored_versions": ["2.7", "3.6", "3.7", "3.9", "3.10"],
```
Let's test on (only) Python 3.9 since that's the latest supported version.

davidcavazos commented on dataflow/extensible-templates/dataflow_udf_transform.js:
```diff
@@ -0,0 +1,35 @@
+ * @param {string} inJson input Pub/Sub JSON message (stringified)
+ */
+ function process(inJson) {
+    var obj = JSON.parse(inJson),
```
It looks like [Nashorn (partially) supports ECMA6](https://openjdk.org/jeps/292), so let's see if we can modernize this Javascript a little bit.

Does it work if we change `var` to `const`? This forces the "variable" to be immutable, and makes it respect naming scopes.

anguillanneuf replied:
No, it doesn't unfortunately. Tests failed after I made the switch. Keeping `var` but we should file a bug.

davidcavazos replied:
Yeah, let's keep `var` for now. Although if Nashorn is getting deprecated in the next Java version, maybe we can let it be. It looks like Java 11 will end of life in around 1 year and alongside with that [Nashorn will be deprecated](https://www.graalvm.org/22.2/reference-manual/js/NashornMigrationGuide/), the templates and other UDFs will most likely have to adapt GraalVM or something similar since Nashorn won't be around.

davidcavazos commented on dataflow/extensible-templates/dataflow_udf_transform.js:
```diff
@@ -0,0 +1,35 @@
+ */
+ function process(inJson) {
+    var obj = JSON.parse(inJson),
+        includePubsubMessage = obj.data && obj.attributes,
```
Nit: I found the use of commas a little confusing to define multiple variables in the same statement since I didn't know that feature. It might be easier to read if we explicitly define each variable in their own statement. I'll leave that to you as personal style, I don't have a strong opinion.

anguillanneuf commented:
Thanks @dandhlee and @davidcavazos for the review. Addressed all the review comments.

For the naming, we agreed on not updating extensible template to classic template because the correct name is actually "Google-provided templates". All "Google-provided templates" are now extensible and we can also create flex templates that are extensible. "extensible" templates aren't really a type of templates defined in the same sense as a feature but a concept.

davidcavazos commented:
LGTM! Thank you! This looks great :)
